### PR TITLE
Introduce snapshot testing

### PR DESCRIPTION
**Pull request by philippfromme**
### Proposed Changes

Allows us to easily compare before and after for the defined test cases. Could be added to the CI so the CI only passes once all snapshots pass meaning you either haven't changed the algorithm or you did and you updated the snapshots.

* snapshot tests are part of `npm test` and a snapshot that is not matching will result in an unsuccessful test run
* snapshots that are matching are highlighted in green
* snapshots that are not matching are highlighted in red
* if no snapshot exists no highlight is added

brave_jawCRplABm

 

### Checklist

To ensure you provided everything we need to look at your PR:

* [x] **Brief textual description** of the changes present
* [x] **Visual demo** attached
* [ ] **Steps to try out** present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [ ] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

### CODE REVIEW

**Comment by nikku** [comment]
Let's get this incorporated soonish, in order to test the other open PRs.

**Comment by philippfromme** [comment]
I'll get this PR ready.

**Comment by barmac** [comment]
@philippfromme looks that we need to make sure our snapshots are initially valid.

**Comment by philippfromme** [comment]
@barmac Thought I'd done that 🙈

File: test/LayoutSpec.js [hunk 1]
```diff
@@ -1,156 +1,104 @@
...
 
-    const generatedContents = generatedDiagrams.map(
-      diagram => fs.readFileSync(outputDirectory + diagram, 'utf8')
-    );
+          // then
+          assert.strictEqual(output, snapshot, `Snapshot does not match output for ${ fileName }`);
+        }
 
-    const generated = generatedContents.map((contents, idx) => {
+        fs.writeFileSync(path.join(outputDirectory, fileName), output, 'utf8');
```
Comment (barmac) [inline test/LayoutSpec.js:47]: Note that the result is not written if the test fails. I'd suggest to write it before the assertion.

File: test/LayoutSpec.js [hunk 2]
```diff
@@ -1,156 +1,104 @@
...
-
-        const containerAfter = document.createElement('div');
-        containerAfter.className = 'container after';
-
-        parent.appendChild(title);
-        parent.appendChild(containerParent);
-
-        containerParent.appendChild(containerBefore);
-        containerParent.appendChild(containerAfter);
+  after(() => {
```
Comment (barmac) [inline test/LayoutSpec.js:56]: Quite unexpected to me to see the teardown code on the bottom. Usual practice is to have tests setup on the top, and place test cases later:

before
after
it

File: test/LayoutSpec.js [hunk 3]
```diff
@@ -1,156 +1,104 @@
...
 
-    const generated = generatedContents.map((contents, idx) => {
+        fs.writeFileSync(path.join(outputDirectory, fileName), output, 'utf8');
 
-      return {
-        diagram: generatedDiagrams[idx],
-        generatedContents: generatedContents[idx],
-        originalContents: originalContents[idx]
-      };
+        if (UPDATE_SNAPSHOTS) {
```
Comment (barmac) [inline test/LayoutSpec.js:49]: Does it work given the assertion above?

File: test/LayoutSpec.js (hunk 1 above)
Comment (philippfromme) [reply test/LayoutSpec.js:47]: Good point!

**Comment by abdul99ahad** [comment]
Some of the test cases are not showing in the left panel. 
image

**Comment by philippfromme** [comment]
> Some of the test cases are not showing in the left panel.

This is because the original diagram doesn't contain any graphical information. While the use case of the library is to create missing graphical information for the snapshot tests it's actually more convinient to have a diagram with graphical information. The library will then simply replace that. I'll update the diagrams that have no DI.

**Comment by philippfromme** [comment]
I updated the diagrams without DI and also added an error indication if the import fails completely. Should be clearer now.

brave_kVlMfV6NoR

**Comment by philippfromme** [comment]
Ready for review.

**Comment by abdul99ahad** [comment]
> Ready for review.

Looks good to me 🚀

**Comment by nikku** [comment]
Updated with a minor hint regarding test updates:

image

**Review by nikku — APPROVED** [review]
Good stuff! I'll look into simplifying our existing tests in a minute.

**Pull request merged**